src/libraries/System.Data.Common/src/System/Data/DataRowExtensions.cs

```diff
@@ -141,8 +149,11 @@ public static void SetField<T>(this DataRow row, DataColumn column, T? value)
 
         private static class UnboxT<T>
         {
+            [UnconditionalSuppressMessage("AOT analysis", "IL3050:RequiresDynamicCode",
```

Comment:
This looks way too complicated for what it does. 

Can you try deleting this and replacing all `UnboxT<T>.s_unbox` with calls to this method:

```csharp
static T Unbox<T>(object o)
{
    if (value == DBNull.Value)
    {
         if (default(T) is null)
             return default(T);
         throw new DataSetUtil.InvalidCast(SR.Format(SR.DataSetLinq_NonNullableCast, typeof(T)));
    }
    return (T)o;
}
```

I think it will do the same thing as this code. No AOT annotations needed.

Reply:
@jkotas the test is failing here basically because this throws "Unable to cast object of type 'System.Int32' to type 'System.Nullable`1[SomeEnum]":

```
object o1 = (int?)0;
Console.WriteLine((int)o1); // ok
Console.WriteLine((int?)o1); // ok
Console.WriteLine((SomeEnum)o1); // ok
Console.WriteLine((SomeEnum?)o1); // throws

enum SomeEnum { }
```

I can't find a type safe way to make this work. The fact we don't allow it to unbox kind of breaks Nullable's main promise:

https://github.com/dotnet/runtime/blob/26b088644417b257906b09f8f62c1e58b6ae0392/src/libraries/System.Private.CoreLib/src/System/Nullable.cs#L11-L13

Can you think of a type safe way to express this?

Do we need a new API? Or should we make the unboxing work? (I could see an argument for the latter but also it's pretty late to do that.)

Reply:
> Can you think of a type safe way to express this?

It can be done via reflection. It is shower than what's there today so we want to take the reflection path only when we have to. Something like:

```csharp
static class UnboxT<T>
{
    internal static readonly Func<object, T?> s_unbox = Create();

    private static Func<object, T?> Create()
    {
        if (typeof(T).IsValueType && default(T) == null)
        {
            if (!RuntimeFeature.IsDynamicCodeSupported)
                return NullableFieldUsingReflection;

            return typeof(UnboxT<T>)
                .GetMethod("NullableField", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic)!
                .MakeGenericMethod(Nullable.GetUnderlyingType(typeof(T))!)
                .CreateDelegate<Func<object, T>>();
        }
        return NonNullableField;
    }

    private static T? NonNullableField(object value)
    {
        if (value == DBNull.Value)
        {
            if (default(T) is null)
                return default;
            throw DataSetUtil.InvalidCast(SR.Format(SR.DataSetLinq_NonNullableCast, typeof(T)));
        }
        return (T)value;
    }

    private static T? NullableFieldUsingReflection(object value)
    {
        if (value == DBNull.Value)
            return default;

        // Try regular cast first
        if (value is T t)
            return t;

        Type valueType = value.GetType();
        Type nullableType = Nullable.GetUnderlyingType(typeof(T))!;

        // Convert does all sorts of conversions. We are only interested in conversions for enums.
        Type fromType = valueType.IsEnum ? Enum.GetUnderlyingType(valueType) : valueType;
        Type toType = nullableType.IsEnum ? Enum.GetUnderlyingType(nullableType) : nullableType;
        if (fromType == toType)
            value = Convert.ChangeType(value, nullableType, null);

        return (T)value;
    }

    private static Nullable<TElem> NullableField<TElem>(object value) where TElem : struct
        => value == DBNull.Value ? default : new Nullable<TElem>((TElem)value);
}
```

> Do we need a new API? Or should we make the unboxing work?

I am not sure whether it makes sense to keep further relaxing the unboxing/casting rules between enums and ints. We may consider doing that, but not for .NET 7 - it is too late as you have pointed out.

I think this is a use-case for https://github.com/dotnet/csharplang/discussions/6308 or something like that.

Reply:
(I just made small edit in the code above that added `Enum.ToObject` call - in case you have already looked at it.)

Reply:
I cannot come up with anything faster - this might do the trick while still being AOT safe, thanks!

It might need a pragma supression to silence the analyzer. The AOT compiler should already strip the IsDynamicCodeSupported codepath.